clang/lib/DPCT/UserDefinedRules/UserDefinedRules.cpp

```diff
@@ -344,17 +344,79 @@ MetaRuleObject::PatternRewriter::PatternRewriter(
   Subrules = S;
 }
 
+// Read a YAML file recursively and substitute any "!include <filename>"
+// directive with the contents of the referenced file.
+std::unique_ptr<llvm::MemoryBuffer>
+readYAMLFile(const llvm::StringRef &ruleFilePath) {
+  // Load the rule file into a MemoryBuffer
+  llvm::ErrorOr<std::unique_ptr<llvm::MemoryBuffer>> Buffer =
+      llvm::MemoryBuffer::getFile(ruleFilePath);
+  if (!Buffer) {
+    llvm::errs() << "Error: failed to read " << ruleFilePath << ": "
+                 << Buffer.getError().message() << "\n";
+    clang::dpct::ShowStatus(MigrationErrorInvalidRuleFilePath);
+    dpctExit(MigrationErrorInvalidRuleFilePath);
+  }
+
+  // Get the directory path of the rule file.
+  llvm::SmallString<128> directoryPath(ruleFilePath);
+  llvm::sys::path::remove_filename(directoryPath);
+
+  // Split the file content into lines using newline as the delimiter.
+  llvm::StringRef buffContent = std::move(*Buffer)->getBuffer();
+  llvm::SmallVector<StringRef, 16> lines;
+  buffContent.split(lines, "\n");
+
+  std::stringstream output;
+  for (auto line : lines) {
+    auto lineStr = line.str();
+
+    // Extract the filename following !include.
+    std::istringstream iss(lineStr);
+    std::string incDirective, incRuleFilePath;
+    // iss >> incDirective >> incRuleFilePath;
+    iss >> incDirective;
+    std::getline(iss >> std::ws, incRuleFilePath);
+
+    // Trim trailing spaces from incRuleFilePath
+    incRuleFilePath.erase(incRuleFilePath.find_last_not_of(" \t\n\r\f\v") + 1);
+
+    if (incDirective == "!include") {
+      if (!incRuleFilePath.empty()) {
+        // Remove surrounding quotes if they exist.
+        if (incRuleFilePath.front() == '"' && incRuleFilePath.back() == '"') {
+          incRuleFilePath =
+              incRuleFilePath.substr(1, incRuleFilePath.size() - 2);
+        }
+
+        // Calculate the absolute path of the included file based on its parent
+        // rule file
+        llvm::SmallString<128> incRuleFileAbsPath = directoryPath;
+        llvm::sys::path::append(incRuleFileAbsPath, incRuleFilePath);
+
+        // Recursively process the included file.
+        if (llvm::sys::fs::exists(incRuleFileAbsPath)) {
+          output << readYAMLFile(incRuleFileAbsPath.str())->getBuffer().str()
+                 << "\n";
+        } else {
+          output << readYAMLFile(incRuleFilePath)->getBuffer().str() << "\n";
+        }
+      } else {
+        continue;
+      }
+    } else {
+      output << lineStr << "\n";
+    }
```

Comment:
I think a single for-loop is enough to handle this.

Reply:
Reimplemented as suggested.
Please review again now.